### ISSUE TITLE
[MJAVADOC-427] "Error fetching URL" for valid non-Java API links

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-javadoc-plugin#729]
**[Selena Renee Phillips](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=selenareneephillips)** opened **[MJAVADOC-427](https://issues.apache.org/jira/browse/MJAVADOC-427?redirect=false)** and commented

When I generate Javadoc for a simple project with 3rd party dependencies, no configuration generates links for these dependencies. For valid Javadoc links which have a package-list, I get the following error:

[WARNING] javadoc: warning - Error fetching URL

Test Cases w/ results (sample project(s)) attached:

**1. No manual link configuration. 'detectLinks' is set to true.**

```
<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-javadoc-plugin</artifactId>
                <version>2.10.3</version>
                <configuration>
                    <reportOutputDirectory>${project.basedir}/target</reportOutputDirectory>
                    <destDir>javadoc</destDir>
                    <windowtitle>Epiphany</windowtitle>
                    <doctitle>Epiphany</doctitle>
                    <show>private</show>
                    <detectLinks>true</detectLinks>
                    <detectOfflineLinks>true</detectOfflineLinks>
                    <linksource>true</linksource>
                    <detectJavaApiLink>true</detectJavaApiLink>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>javadoc</goal>
                            <goal>test-javadoc</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

Output of 'mvn javadoc:javadoc':

```
[INFO] Scanning for projects...
[INFO]
[INFO] Using the builder org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder with a thread count of 1
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building epiphany 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] >>> maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany >>>
[INFO]
[INFO] <<< maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany <<<

....SNIP DOWNLOADING INFO.....

[INFO]
[INFO] --- maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany ---
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-java/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-api/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-support/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://code.google.com/p/guava-libraries/guava/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://logback.qos.ch/logback-core/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://logback.qos.ch/logback-classic/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://testng.org/apidocs/package-list. Ignored it.
[INFO]
Loading source files for package com.example...
Constructing Javadoc information...
Standard Doclet version 1.8.0_05
Building tree for all the packages and classes...
Generating C:\Users\sephilli\git\javadoc-bug\target\javadoc\com\example\AbstractLoadable.html...

.....SNIP....

1 warning
[WARNING] Javadoc Warnings
[WARNING] javadoc: warning - Error fetching URL: http://www.slf4j.org/apidocs
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 13.694 s
[INFO] Finished at: 2015-04-18T08:24:05-05:00
[INFO] Final Memory: 16M/92M
[INFO] ------------------------------------------------------------------------
```

I expected that the plugin would fail to find javadoc to link for Selenium, Logback, Guava and TestNg, but to succeed in fetching Javadoc for SLF4Jj. The detected link for SLF4J is a valid javadoc link with a package-list located at the detected link:

http://www.slf4j.org/apidocs/
http://www.slf4j.org/apidocs/package-list

However, there is a error fetching the URL and the resulting javadoc does not have links for SLF4J:

```
log
private static final org.slf4j.ext.XLogger log
The logger for the AbstractLoadable class
```

**2. Manual link configuration for dependencies with javadoc that is not automatically detected with 'detectLinks' is set to true. No trailing slashes in manually configured links.**

```
<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-javadoc-plugin</artifactId>
                <version>2.10.3</version>
                <configuration>
                    <reportOutputDirectory>${project.basedir}/target</reportOutputDirectory>
                    <destDir>javadoc</destDir>
                    <windowtitle>Epiphany</windowtitle>
                    <doctitle>Epiphany</doctitle>
                    <show>private</show>
                    <detectLinks>true</detectLinks>
                    <detectOfflineLinks>true</detectOfflineLinks>
                    <linksource>true</linksource>
                    <detectJavaApiLink>true</detectJavaApiLink>
                    <links>
                        <link>https://selenium.googlecode.com/git/docs/api/java</link>
                        <link>http://testng.org/javadocs</link>
                        <link>http://logback.qos.ch/apidocs</link>
                        <link>http://docs.guava-libraries.googlecode.com/git/javadoc</link>                       
                    </links>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>javadoc</goal>
                            <goal>test-javadoc</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

Output from mvn javadoc:javadoc:

```
[INFO] Scanning for projects...
[INFO]
[INFO] Using the builder org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder with a thread count of 1
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building epiphany 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] >>> maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany >>>
[INFO]
[INFO] <<< maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany <<<
....SNIP DOWNLOADING INFO ....
[INFO]
[INFO] --- maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany ---
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-java/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-api/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-support/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://code.google.com/p/guava-libraries/guava/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://logback.qos.ch/logback-core/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://logback.qos.ch/logback-classic/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://testng.org/apidocs/package-list. Ignored it.
[INFO]
Loading source files for package com.example...
Constructing Javadoc information...
Standard Doclet version 1.8.0_05
Building tree for all the packages and classes...
Generating C:\Users\sephilli\git\javadoc-bug\target\javadoc\com\example\AbstractLoadable.html...
.....SNIP......
5 warnings
[WARNING] Javadoc Warnings
[WARNING] javadoc: warning - Error fetching URL: https://selenium.googlecode.com/git/docs/api/java
[WARNING] javadoc: warning - Error fetching URL: http://testng.org/javadocs
[WARNING] javadoc: warning - Error fetching URL: http://logback.qos.ch/apidocs
[WARNING] javadoc: warning - Error fetching URL: http://docs.guava-libraries.googlecode.com/git/javadoc
[WARNING] javadoc: warning - Error fetching URL: http://www.slf4j.org/apidocs
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:18 min
[INFO] Finished at: 2015-04-18T09:05:08-05:00
[INFO] Final Memory: 16M/95M
[INFO] ------------------------------------------------------------------------

```

The manually configured links are all valid javadoc links with package-list present at that location, but resulting javadoc does not link third-party doc for dependencies:

```
    driver
    @Nonnull
    private final org.openqa.selenium.WebDriver driver

    The driver for browsing this AbstractLoadable
```

**3. Manual link configuration for dependencies with javadoc that is not automatically detected with 'detectLinks' is set to true. Trailing slashes in manually configured links**

```
<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-javadoc-plugin</artifactId>
                <version>2.10.3</version>
                <configuration>
                    <reportOutputDirectory>${project.basedir}/target</reportOutputDirectory>
                    <destDir>javadoc</destDir>
                    <windowtitle>Epiphany</windowtitle>
                    <doctitle>Epiphany</doctitle>
                    <show>private</show>
                    <detectLinks>true</detectLinks>
                    <detectOfflineLinks>true</detectOfflineLinks>
                    <linksource>true</linksource>
                    <detectJavaApiLink>true</detectJavaApiLink>
                    <links>
                        <link>https://selenium.googlecode.com/git/docs/api/java/</link>
                        <link>http://testng.org/javadocs/</link>
                        <link>http://logback.qos.ch/apidocs/</link>
                        <link>http://docs.guava-libraries.googlecode.com/git/javadoc/</link>                       
                    </links>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>javadoc</goal>
                            <goal>test-javadoc</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

Output from mvn javadoc:javadoc:

```
[INFO] Scanning for projects...
[INFO]
[INFO] Using the builder org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder with a thread count of 1
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building epiphany 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] >>> maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany >>>
[INFO]
[INFO] <<< maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany <<<
...SNIP DOWNLOADING INFO ...
[INFO]
[INFO] --- maven-javadoc-plugin:2.10.3:javadoc (default-cli) @ epiphany ---
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-java/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-api/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://selenium.googlecode.com/selenium-support/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://code.google.com/p/guava-libraries/guava/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://logback.qos.ch/logback-core/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://logback.qos.ch/logback-classic/apidocs/package-list. Ignored it.
[ERROR] Error fetching link: http://testng.org/apidocs/package-list. Ignored it.
[INFO]
Loading source files for package com.example...
Constructing Javadoc information...
Standard Doclet version 1.8.0_05
Building tree for all the packages and classes...
Generating C:\Users\sephilli\git\javadoc-bug\target\javadoc\com\example\AbstractLoadable.html...
....SNIP....
5 warnings
[WARNING] Javadoc Warnings
[WARNING] javadoc: warning - Error fetching URL: https://selenium.googlecode.com/git/docs/api/java
[WARNING] javadoc: warning - Error fetching URL: http://testng.org/javadocs
[WARNING] javadoc: warning - Error fetching URL: http://logback.qos.ch/apidocs
[WARNING] javadoc: warning - Error fetching URL: http://docs.guava-libraries.googlecode.com/git/javadoc
[WARNING] javadoc: warning - Error fetching URL: http://www.slf4j.org/apidocs
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:18 min
[INFO] Finished at: 2015-04-18T09:20:17-05:00
[INFO] Final Memory: 16M/110M
[INFO] ------------------------------------------------------------------------
```

Resulting javadoc does not link the javadoc for the 3rd party dependencies.

**Test Case 4: detectLinks= false + manually configured links without trailing slashes**

Result is the same. I won't post the pom and the output here because the sample project(s) are attached and the result is not different from the previous test cases. Errors are reporting for attempts to fetch all of the URLs.

**Test Case 5: detectLinks-false + manually configured links with trailing slashes.**

Result is the same.

---

**Affects:** 2.10.3

**Attachments:**
- [testcase-detectLinks-enabled-no-manual-link-config.zip](https://issues.apache.org/jira/secure/attachment/12726361/testcase-detectLinks-enabled-no-manual-link-config.zip) (_11.02 kB_)
- [testcase-detectLinks-enabled-with-manual-link-config-no-trailing-slash.zip](https://issues.apache.org/jira/secure/attachment/12726362/testcase-detectLinks-enabled-with-manual-link-config-no-trailing-slash.zip) (_11.64 kB_)
- [testcase-detectLinks-enabled-with-manual-link-config-trailing-slash.zip](https://issues.apache.org/jira/secure/attachment/12726363/testcase-detectLinks-enabled-with-manual-link-config-trailing-slash.zip) (_11.56 kB_)
- [testcases-detectLinks-disabled-with-manual-link-config-no-trailing-slash.zip](https://issues.apache.org/jira/secure/attachment/12726364/testcases-detectLinks-disabled-with-manual-link-config-no-trailing-slash.zip) (_11.71 kB_)
- [testcases-detectLinks-disabled-with-manual-link-config-trailing-slash.zip](https://issues.apache.org/jira/secure/attachment/12726365/testcases-detectLinks-disabled-with-manual-link-config-trailing-slash.zip) (_11.63 kB_)

**Issue Links:**
- [IGNITE-7643](https://issues.apache.org/jira/browse/IGNITE-7643) Broken javadoc in partitioned dataset

- [MDEP-727](https://issues.apache.org/jira/browse/MDEP-727) Broken JavaDoc links

- [MJAVADOC-487](https://issues.apache.org/jira/browse/MJAVADOC-487) Follow redirects

- [MJAVADOC-539](https://issues.apache.org/jira/browse/MJAVADOC-539) Upgrading plugin from 3.0.0 to 3.0.1 and 3.1.0 breaks certain external javadoc links
 (_**"breaks"**_)

17 votes, 19 watchers

**Comment by jira-importer** [comment]
**[Nick Radov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=nick.radov)** commented

This error appears to have first appeared in version 2.7.

**Comment by jira-importer** [comment]
**[Dick Adams](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dadams07)** commented

Bug still present in plugin version 2.10.4

**Comment by jira-importer** [comment]
**[Petr Ivanov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vveider)** commented

Bug still presents in plugin version `3.0.0`.

**Comment by jira-importer** [comment]
**[Guillaume Boué](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gboue)** commented

Looks like the issue is the HTTP -> HTTPS redirection that does not happen inside the `javadoc` tool itself. Possible upstream JDK issue here: [JDK-8190312](https://bugs.openjdk.java.net/browse/JDK-8190312). The plugin instructs `javadoc` to locate the apidocs at `http://www.slf4j.org/apidocs` (because slf4j URL [is `http://www.slf4j.org`](https://github.com/qos-ch/slf4j/blob/v_1.7.12/pom.xml#L15) in its POM) and then it does not follow the redirect to HTTPS.

Indeed, if we explicitly add `<link>https://www.slf4j.org/apidocs</link>`, the link gets correctly generated.

**Comment by jira-importer** [comment]
**[Matt Nelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mattnelson)** commented

Logged MJAVADOC-487 for this same issue.
 
Can JavadocUtil[1] and AbstractJavadocMojo[2] be enhanced to handle the redirect before handing the URI off to the javadoc tool?

[1] https://github.com/apache/maven-javadoc-plugin/blob/12dbbde29cf6277ca311cb8afffdf02dbfe0c9b4/src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java#L1789
[2] https://github.com/apache/maven-javadoc-plugin/blob/12dbbde29cf6277ca311cb8afffdf02dbfe0c9b4/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java#L5883

**Comment by jira-importer** [comment]
**[Guillaume Boué](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gboue)** commented

Yes, since `javadoc` currently doesn't follow redirects, we can do it in the plugin and then feed `javadoc` the last redirect location. I'll work on this.

**Comment by jira-importer** [comment]
**[Guillaume Boué](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gboue)** commented

Fixed in [0ade0576e8a294459153161a3d5071814e588dcd](https://gitbox.apache.org/repos/asf?p=maven-javadoc-plugin.git;a=commit;h=0ade0576e8a294459153161a3d5071814e588dcd).
